lib/Service/ParticipantService.php

```diff
@@ -300,6 +307,66 @@ public function addUsers(Room $room, array $participants): void {
 		$this->dispatcher->dispatch(Room::EVENT_AFTER_USERS_ADD, $event);
 	}
 
+	/**
+	 * @param Room $room
+	 * @param IGroup $group
+	 * @param Participant[] $existingParticipants
+	 */
+	public function addGroup(Room $room, IGroup $group, array $existingParticipants = []): void {
+		$usersInGroup = $group->getUsers();
+
+		if (empty($existingParticipants)) {
+			$existingParticipants = $this->getParticipantsForRoom($room);
+		}
+
+		$participantsByUserId = [];
+		foreach ($existingParticipants as $participant) {
+			if ($participant->getAttendee()->getActorType() === Attendee::ACTOR_USERS) {
+				$participantsByUserId[$participant->getAttendee()->getActorId()] = $participant;
+			}
+		}
+
+		$newParticipants = [];
+		foreach ($usersInGroup as $user) {
+			$existingParticipant = $participantsByUserId[$user->getUID()] ?? null;
+			if ($existingParticipant instanceof Participant) {
+				if ($existingParticipant->getAttendee()->getParticipantType() === Participant::USER_SELF_JOINED) {
+					$this->updateParticipantType($room, $existingParticipant, Participant::USER);
+				}
+
+				// Participant is already in the conversation, so skip them.
+				continue;
+			}
+
+			$newParticipants[] = [
+				'actorType' => Attendee::ACTOR_USERS,
+				'actorId' => $user->getUID(),
+				'displayName' => $user->getDisplayName(),
+			];
+		}
+
+		try {
+			$this->attendeeMapper->findByActor($room->getId(), Attendee::ACTOR_GROUPS, $group->getGID());
+		} catch (DoesNotExistException $e) {
+			$lastMessage = 0;
+			if ($room->getLastMessage() instanceof IComment) {
+				$lastMessage = (int) $room->getLastMessage()->getId();
+			}
+
+			$attendee = new Attendee();
+			$attendee->setRoomId($room->getId());
+			$attendee->setActorType(Attendee::ACTOR_GROUPS);
+			$attendee->setActorId($group->getGID());
+			$attendee->setDisplayName($group->getDisplayName());
+			$attendee->setParticipantType(Participant::USER);
+			$attendee->setLastReadMessage($lastMessage);
+			$attendee->setReadPrivacy(Participant::PRIVACY_PUBLIC);
```

Comment:
Why is it needed (or, maybe, advisable) to set the last read message or the read privacy for the group?

Reply:
Right, privacy needs to be set to private so the read status is ignored.
Last read message I set out of random.

Reply:
FYI I tried with a public privacy, a private privacy and not explicitly setting any privacy for the group and in all cases having a group in the conversation did not seem to affect whether a message was seen as read or not (it changed only if the other participants actually read it). So I do not know if I made something wrong, it is actually not needed or there is some bug lurking around in the read status :shrug:

Reply:
https://github.com/nextcloud/spreed/blob/2183ffc24a5abdbcd81473422c14da74d87af941/lib/Service/ParticipantService.php#L542 we also check the user type, otherwise an email invite would break it already :D I just forgot about it